src/queue.rs

```diff
@@ -0,0 +1,148 @@
+                active_model.status = Set(JobStatus::Success);
+                active_model.scheduled_at = Set(None);
+                let mut next_job = ActiveModel::from(job);
+                next_job.parent_id = Set(Some(*active_model.id.as_ref()));
+                let next_job = next_job.insert(&tx).await?;
+                active_model.result = Set(Some(JobResult::Child(next_job.id)));
+            }
+
+            Ok(None) => {
+                active_model.scheduled_at = Set(None);
+                active_model.status = Set(JobStatus::Success);
+                active_model.result = Set(Some(JobResult::Complete));
+            }
+
+            Err(e) if e.is_retryable() => {
+                active_model.failure_count = Set(active_model.failure_count.as_ref() + 1);
+                let reschedule =
+                    schedule_based_on_failure_count(*active_model.failure_count.as_ref());
+                active_model.status =
+                    Set(reschedule.map_or(JobStatus::Failed, |_| JobStatus::Pending));
+                active_model.scheduled_at = Set(reschedule);
+                active_model.result = Set(Some(JobResult::Error(e)));
+            }
+
+            Err(e) => {
+                active_model.failure_count = Set(active_model.failure_count.as_ref() + 1);
+                active_model.scheduled_at = Set(None);
+                active_model.status = Set(JobStatus::Failed);
+                active_model.result = Set(Some(JobResult::Error(e)));
+            }
+        }
+        Some(active_model.update(&tx).await?)
+    } else {
+        None
+    };
+    tx.commit().await?;
+    Ok(model)
+}
+
+fn spawn(join_set: &mut JoinSet<()>, db: &Db, job_state: &Arc<SharedJobState>) {
+    let db = db.clone();
+    let job_state = Arc::clone(job_state);
+    join_set.spawn(async move {
+        loop {
+            match one(&db, &job_state).await {
+                Err(e) => {
+                    eprintln!("job error {e}");
+                }
+
+                Ok(Some(_)) => {}
+
+                Ok(None) => {
+                    sleep(Duration::from_secs(fastrand::u64(QUEUE_CHECK_INTERVAL))).await;
+                }
+            }
+        }
+    });
+}
+
+pub async fn run(db: Db, config: ApiConfig) {
+    let mut join_set = JoinSet::new();
+    let job_state = Arc::new(SharedJobState::from(&config));
+    for _ in 0..QUEUE_WORKER_COUNT {
+        spawn(&mut join_set, &db, &job_state);
+    }
+
+    while join_set.join_next().await.is_some() {
+        eprintln!("Worker task shut down. Restarting.");
+        spawn(&mut join_set, &db, &job_state);
+    }
+}
+
+async fn next(tx: &DatabaseTransaction) -> Result<Option<Model>, DbErr> {
+    let select = match tx.get_database_backend() {
+        backend @ DatabaseBackend::Postgres => Statement::from_sql_and_values(
+            backend,
+            r#"SELECT * FROM queue
+               WHERE status = $1 AND (scheduled_at IS NULL OR scheduled_at < $2)
```

Comment:
This queue strategy works by holding a DB lock on the relevant row for the length of processing of the queue item, so e.g. for sending email it would hold a DB lock across the API call to Postmark.

This probably works for the scale we envision. But we might want to set an appropriately-long `scheduled_at` value, process the record outside of a transaction (with a timeout based on `scheduled_at`), then update status in a second transaction. (This is effectively a "lease" over the queue item.)

Reply:
Is the concern here that we will exhaust the connection pool?

Reply:
Indeed, or that too-long transactions may lead to locking contention. (I think the touches to the queue table itself are unlikely to contend, as not contending is the entire purpose of the way the queries are written; but queries made while running the job itself might.)

Actually, now that I think about it, what isolation level are we running under for Postgres? I would hope for at least Repeatable Read; but if we're running at that level, a serialization failure at the very end of the transaction might lead to retrying processing of the job, which in practice would send a duplicate email. I think in that case, we need to find some way to avoid transaction retries necessitating retries of job-processing logic; one way to do so would be the suggested "lease" system, though there are other solutions too (we had previously discussed the queue states as being `START`, `IN_PROCESS`, `DONE`, with the `IN_PROCESS` state meaning "currently processing, don't pick me up, if I crash we get stuck here).